source/agora/common/crypto/Schnorr.d

```diff
@@ -269,3 +269,80 @@ nothrow @nogc @safe unittest
     assert(verify(kp2.V, sig2, secret));
     assert(!verify(kp2.V, sig1, secret));
 }
+
+// example of extracting the private key from an insecure signature scheme
+// which did not include 'r' during the signing
+// https://tlu.tarilabs.com/cryptography/digital_signatures/introduction_schnorr_signatures.html#why-do-we-need-the-nonce
+/*@nogc*/ @safe unittest
+{
+    static immutable string message = "BOSAGORA for the win";
+
+    Pair kp = Pair.random();  // key-pair
+    Scalar c = hashFull(message);  // challenge
+    Scalar s = (kp.v * c);  // signature
+
+    // known public data of the node
+    Point K = kp.V;
+
+    // other nodes verify
+    assert(s.toPoint() == K * c);
+
+    // but the other node can also extract the private key!
+    Scalar stolen_key = s * c.invert();
+    assert(stolen_key == kp.v);
+}
+
+// possibly secure signature scheme
+/*@nogc*/ @safe unittest
+{
+    static immutable string message = "BOSAGORA for the win";
+
+    Pair kp = Pair.random();  // key-pair
+    Pair Rp = Pair.random();  // (R, r), the public and private nonce
+    Scalar c = hashFull(message);  // challenge
+    Scalar s = Rp.v + (kp.v * c);  // signature
+
+    // known public data of the node
+    Point K = kp.V;
+    Point R = Rp.V;
+
+    // other nodes verify
+    assert(s.toPoint() == R + (K * c));
+
+    // other nodes cannot extract the private key, they don't know 'r'
+    Scalar stolen_key = s * c.invert();
+    assert(stolen_key != kp.v);
+}
+
+// possibly secure signature scheme, demonstrating arbitrary addition of signatures
+/*@nogc*/ @safe unittest
+{
+    static immutable string message = "BOSAGORA for the win";
+
+    Scalar c = hashFull(message);  // challenge
+
+    Pair kp_1 = Pair.random();  // key-pair
+    Pair Rp_1 = Pair.random();  // (R, r), the public and private nonce
+    Scalar s_1 = Rp_1.v + (kp_1.v * c);  // signature
+
+    Pair kp_2 = Pair.random();  // key-pair
+    Pair Rp_2 = Pair.random();  // (R, r), the public and private nonce
+    Scalar s_2 = Rp_2.v + (kp_2.v * c);  // signature
+
+    // known public data of the nodes
+    Point K_1 = kp_1.V;
+    Point R_1 = Rp_1.V;
+
+    Point K_2 = kp_2.V;
+    Point R_2 = Rp_2.V;
+
+    // verification of individual signatures
+    assert(s_1.toPoint() == R_1 + (K_1 * c));
+    assert(s_2.toPoint() == R_2 + (K_2 * c));
+
+    // "multi-sig" - collection of one or more signatures
+    auto sum_s = s_1 + s_2;
+    assert(sum_s.toPoint() ==
+        (R_1 + (K_1 * c)) +
+        (R_2 + (K_2 * c)));
```

Comment:
```suggestion
    assert(sum_s.toPoint() ==
        (R_1 + (K_1 * c)) +
        (R_2 + (K_2 * c)));
    // Or the equivalent:
    assert(sum_s.toPoint() ==
        (R_1 + R2) + (K_1 * c) + (K_2 * c));
```

Reply:
The reason for this suggestion is that it will be closer to what's done during signature aggregation.

Reply:
I'm not sure I follow but I've applied it regardless.